File: java/main/java/src/kata/two/SecondKata.java
package kata.two;

/**
 * Created by Eureka.
 * User: ggarcia
 * Date: 4/26/11
 * Time: 15:28
 *
 * @author ggarcia
 */
public class SecondKata {
    public long sum(double limit) {
        long n = Math.round(((Math.log(limit * Math.sqrt(5))) * 2) + 1.618);
        long sum = 0;

        for (long i = 6; i < n; i++) {
            long fib = (4 * fib(i - 3)) + fib(i - 6);

            if (fib % 2 == 0)
                sum += fib;
        }

        return sum + 2;
    }

    private long fib(long n) {
        return (n < 2) ? n : fib(n - 1) + fib(n - 2);
    }
}

File: scala/main/scala/src/katas/two/SecondKata.scala
package katas.two

import scala.math._

/**
 * Created by Eureka.
 * User: ggarcia
 * Date: 5/09/11
 * Time: 14:26
 *
 * @author ggarcia
 */
class SecondKata {
  def sum(limit: Long): Long = {
    val n: Long = round(((log(limit * sqrt(5))) * 2) + 1.618)

    var sum: Long = 0

    (for (i: Long <- 6l to 40l;
          j = (4 * fib(i - 3)) + fib(i - 6);
          if (j < limit) && isEven(j)
    ) yield j).foreach(sum += _)

    sum + 2
  }

  def isEven(n: Long): Boolean = {
    return (n % 2) == 0
  }

  def fib(n: Long): Long = {
    if ((n < 2)) n else fib(n - 1) + fib(n - 2)
  }
}